File: python/cote/LTCD/페어의_노드_스왑.py
import random

TEST_CASES = 10
MAX_LENGTH = 10
MAX_VALUE = 1000


class Node:
    def __init__(self, value, next_=None):
        self.value = value
        self.next = next_

    def __iter__(self):
        cur = self
        while cur != None:
            yield cur.value
            cur = cur.next


def solve(node: Node):
    root = prev = Node(None)

    prev.next = node

    while node and node.next:

        b = node.next
        node.next = b.next
        b.next = node
        prev.next = b

        node = node.next
        prev = prev.next.next

    return root.next


def solve2(node: Node):

    if node and node.next:
        b = node.next

        node.next = solve2(b.next)
        b.next = node
        return b

    return node


if __name__ == "__main__":
    random.seed(43)

    for _ in range(TEST_CASES):

        n = random.randrange(1, MAX_LENGTH)

        prev = Node(random.randrange(MAX_VALUE))

        for _ in range(n - 1):
            prev = Node(random.randrange(MAX_VALUE), prev)

        print([*solve(prev)])

File: go/cote/LTCD/페어의_노드_스왑.go
package LTCD

type 페어의_노드_스왑 struct {
}

type Node5 struct {
	next  *Node5
	value int
}

func (problem 페어의_노드_스왑) makeLinkedList(ints []int) *Node5 {
	node := new(Node5)
	cur := node

	for _, val := range ints {
		cur.next = &Node5{nil, val}
		cur = cur.next
	}

	return node.next
}

func (problem 페어의_노드_스왑) toInts(root *Node5) []int {
	result := make([]int, 0)
	cur := root

	for {
		if cur == nil {
			break
		}

		result = append(result, cur.value)

		cur = cur.next
	}

	return result
}

func (problem 페어의_노드_스왑) solve(root *Node5) *Node5 {
	newRoot := &Node5{next: root}
	prev := newRoot

	for {
		if prev.next == nil || prev.next.next == nil {
			break
		}

		a := prev.next
		b := a.next
		c := b.next

		prev.next = b
		a.next = c
		b.next = a

		prev = prev.next.next
	}

	return newRoot.next
}

func (problem 페어의_노드_스왑) solve2(cur *Node5) *Node5 {
	if cur == nil || cur.next == nil {
		return cur
	}

	node := cur.next
	cur.next = problem.solve2(node.next)
	node.next = cur

	return node
}
